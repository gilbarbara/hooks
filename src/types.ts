import { RefObject } from 'react';

export type PlainObject<T = unknown> = Record<PropertyKey, T>;

export type Primitive = bigint | boolean | null | number | string | symbol | undefined;

export type Target<T extends Element> = RefObject<T> | T | null | string;

export type TimerStatus = 'pending' | 'completed' | 'cancelled';
