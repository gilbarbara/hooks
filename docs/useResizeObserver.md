# useResizeObserver

Detect changes in an Element dimensions using the [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) API.  
Returns a `ResizeObserverEntry`.

## Usage

```tsx
import React from 'react';
import { useResizeObserver } from '@gilbarbara/hooks';

function Component() {
  const ref = React.useRef<HTMLDivElement>(null);
  const entry = useResizeObserver(ref);

  console.log(entry?.contentRect.width);

  return <div ref={ref}>Some content...</div>;
}
```

## Reference

```typescript
useResizeObserver<T extends Element>(
  target: RefObject<T> | T | null | string,
  debounce: number = 0,
): ResizeObserverEntry
```

> This hook uses a [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver), so if you want to support
legacy browsers, install the [resize-observer-polyfill](https://www.npmjs.com/package/resize-observer-polyfill) in your app.
