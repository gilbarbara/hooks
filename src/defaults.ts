import type { UseMeasureResult } from './useMeasure';

export const defaultElementDimensions: UseMeasureResult = {
  absoluteHeight: 0,
  absoluteWidth: 0,
  bottom: 0,
  height: 0,
  left: 0,
  right: 0,
  top: 0,
  width: 0,
  x: 0,
  y: 0,
};
