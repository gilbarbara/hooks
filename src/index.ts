export { useClickOutside } from './useClickOutside';
export { useElementSize } from './useElementSize';
export { useFetch, FetchStatus } from './useFetch';
export { useIntersectionObserver } from './useIntersectionObserver';
export { useIsomorphicLayoutEffect } from './useIsomorphicLayoutEffect';
export { useLatest } from './useLatest';
export { useMediaQuery } from './useMediaQuery';
export { useRenderCount } from './useRenderCount';
export { useResize } from './useResize';
export { useResizeObserver } from './useResizeObserver';
export { useResponsive } from './useResponsive';
export { useScript } from './useScript';
export { useSingleton } from './useSingleton';
export { useThrottle } from './useThrottle';
export { useThrottleValue } from './useThrottleValue';
export { useWhyDidYouUpdate } from './useWhyDidYouUpdate';
