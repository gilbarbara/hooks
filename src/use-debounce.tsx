import { useEffect, useState } from 'react';

export default function useDebounce(value: unknown, ms: number): unknown {
  // State and setters for debounced value
  const [debouncedValue, setDebouncedValue] = useState(value);

  useEffect(
    () => {
      // Update debounced value after delay
      const handler = setTimeout(() => {
        setDebouncedValue(value);
      }, ms);

      // Cancel the timeout if value changes (also on delay change or unmount)
      // This is how we prevent debounced value from updating if value is changed ...
      // .. within the delay period. Timeout gets cleared and restarted.
      return () => {
        clearTimeout(handler);
      };
    },
    [value, ms], // Only re-call effect if value or delay changes
  );

  return debouncedValue;
}
