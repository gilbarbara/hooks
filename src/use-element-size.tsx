import { useEffect, useLayoutEffect, useMemo, useState } from 'react';

export type UseElementRect = Omit<DOMRectReadOnly, 'toJSON'>;

const defaultState: UseElementRect = {
  x: 0,
  y: 0,
  width: 0,
  height: 0,
  top: 0,
  left: 0,
  bottom: 0,
  right: 0,
};

export default function useElementSize(selector: string): UseElementRect {
  const [element, setElement] = useState<Element | null>();
  const [rect, setRect] = useState<UseElementRect>(defaultState);

  useEffect(() => {
    setElement(document.querySelector(selector));
  }, [selector]);

  const observer = useMemo(
    () =>
      new window.ResizeObserver(entries => {
        if (entries[0]) {
          const { bottom, height, left, right, top, width, x, y } = entries[0].contentRect;

          setRect({ x, y, width, height, top, left, bottom, right });
        }
      }),
    [],
  );

  useLayoutEffect(() => {
    if (!element) {
      return;
    }

    observer.observe(element);

    // eslint-disable-next-line consistent-return
    return () => {
      observer.disconnect();
    };
  }, [observer, element]);

  return rect;
}
