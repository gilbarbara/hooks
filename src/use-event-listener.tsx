import { useEffect, useRef, MutableRefObject } from 'react';

type Handler = (e: Event) => void;

export default function useEventListener(
  eventName: string,
  handler: Handler,
  element: Window | Document | HTMLElement = window,
) {
  const savedHandler: MutableRefObject<null | Handler> = useRef(null);

  useEffect(() => {
    savedHandler.current = handler;
  }, [handler]);

  useEffect(() => {
    const { current } = savedHandler;
    const eventListener = (event: Event) => {
      /* istanbul ignore else */
      if (current) {
        current!(event);
      }
    };

    element.addEventListener(eventName, eventListener);

    return () => {
      element.removeEventListener(eventName, eventListener);
    };
  }, [element, eventName]);
}
