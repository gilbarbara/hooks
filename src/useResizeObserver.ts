import { useMemo, useRef, useState } from 'react';

import { Target } from './types';
import { useIsomorphicLayoutEffect } from './useIsomorphicLayoutEffect';
import { canUseDOM, getElement } from './utils';

export function useResizeObserver<T extends Element>(target: Target<T>, debounce: number = 0) {
  const [element, setElement] = useState(getElement(target));
  const [value, setValue] = useState<ResizeObserverEntry>();
  const timeoutRef = useRef<number | null>(null);

  const isFirstCall = useRef(true);

  const observer = useMemo(() => {
    if (!canUseDOM()) {
      return {};
    }

    return new window.ResizeObserver(entries => {
      if (debounce && !isFirstCall.current) {
        if (timeoutRef.current) {
          window.clearTimeout(timeoutRef.current);
        }

        timeoutRef.current = window.setTimeout(() => {
          setValue(entries[0]);
        }, debounce);

        return;
      }

      setValue(entries[0]);
      isFirstCall.current = false;
    });
  }, [debounce]);

  useIsomorphicLayoutEffect(() => {
    setElement(getElement(target));
  }, [target]);

  useIsomorphicLayoutEffect(() => {
    if (!canUseDOM() || !(observer instanceof ResizeObserver)) {
      return () => undefined;
    }

    if (!element) {
      return () => undefined;
    }

    observer.observe(element);

    return () => {
      observer.disconnect();
    };
  }, [element, observer]);

  return value;
}
