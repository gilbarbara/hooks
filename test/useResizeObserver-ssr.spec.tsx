/**
 * @jest-environment node
 */

import * as React from 'react';
import { renderToString } from 'react-dom/server';

import { useResizeObserver } from '../src/useResizeObserver';

function Component() {
  useResizeObserver('#root');

  return <div />;
}

describe('useResizeObserver-ssr', () => {
  it('should work without a DOM', () => {
    const view = renderToString(<Component />);

    expect(view).toMatchSnapshot();
  });
});
