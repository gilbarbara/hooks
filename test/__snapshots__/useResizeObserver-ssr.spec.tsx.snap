// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useResizeObserver-ssr should work without a DOM 1`] = `
<div>
</div>
`;
